planning/autoware_path_generator/src/node.cpp

```diff
@@ -0,0 +1,418 @@
+
+  if (!path) {
+    RCLCPP_ERROR_THROTTLE(get_logger(), *get_clock(), 5000, "output path is invalid");
+    return std::nullopt;
+  } else if (path->points.empty()) {
+    RCLCPP_ERROR_THROTTLE(get_logger(), *get_clock(), 5000, "output path is empty");
+    return std::nullopt;
+  }
+
+  return path;
+}
+
+std::optional<PathWithLaneId> PathGenerator::generate_path(
+  const geometry_msgs::msg::Pose & current_pose, const Params & params) const
+{
+  lanelet::ConstLanelet current_lane;
+  if (!lanelet::utils::query::getClosestLanelet(
+        planner_data_.preferred_lanelets, current_pose, &current_lane)) {
+    return std::nullopt;
+  }
+
+  const auto lanelet_sequence = utils::get_lanelet_sequence(
+    current_lane, planner_data_, current_pose, params.forward_path_length,
+    params.backward_path_length);
+  if (!lanelet_sequence) {
+    return std::nullopt;
+  }
+
+  return generate_path(*lanelet_sequence, current_pose, params);
+}
+
+std::optional<PathWithLaneId> PathGenerator::generate_path(
+  const lanelet::LaneletSequence & lanelet_sequence, const geometry_msgs::msg::Pose & current_pose,
+  const Params & params) const
+{
+  if (lanelet_sequence.empty()) {
+    return std::nullopt;
+  }
+
+  const auto arc_coordinates =
+    lanelet::utils::getArcCoordinates(lanelet_sequence.lanelets(), current_pose);
+  const auto s = arc_coordinates.length;  // s denotes longitudinal position in Frenet coordinates
+  const auto s_start = std::max(0., s - params.backward_path_length);
+  const auto s_end = [&]() {
+    auto s_end = s + params.forward_path_length;
+
+    if (!utils::get_next_lanelet_within_route(lanelet_sequence.lanelets().back(), planner_data_)) {
+      const double lane_length = lanelet::geometry::length(lanelet_sequence.centerline2d());
+      s_end = std::min(s_end, lane_length);
+    }
+
+    if (std::any_of(
+          planner_data_.goal_lanelets.begin(), planner_data_.goal_lanelets.end(),
+          [&](const auto & goal_lanelet) {
+            return lanelet_sequence.lanelets().back().id() == goal_lanelet.id();
+          })) {
+      const auto goal_arc_coordinates =
+        lanelet::utils::getArcCoordinates(lanelet_sequence.lanelets(), planner_data_.goal_pose);
+      s_end = std::min(s_end, goal_arc_coordinates.length);
+    }
+
+    return s_end;
+  }();
+
+  auto path = generate_path(lanelet_sequence, s_start, s_end, params);
+  if (!path) {
+    return std::nullopt;
+  }
+
+  path = autoware::motion_utils::resamplePath(
+    *path, params.input_path_interval, params.enable_akima_spline_first);
+
+  const auto current_seg_idx =
+    autoware::motion_utils::findFirstNearestSegmentIndexWithSoftConstraints(
+      path->points, current_pose, params.ego_nearest_dist_threshold,
+      params.ego_nearest_yaw_threshold);
+
+  path->points = autoware::motion_utils::cropPoints(
+    path->points, current_pose.position, current_seg_idx, params.forward_path_length,
+    params.backward_path_length + params.input_path_interval);
```

Comment:
will be replaced with Autoware Trajectory
https://autowarefoundation.github.io/autoware.universe/main/common/autoware_trajectory/